presto-hive/src/main/java/com/facebook/presto/hive/rule/BaseHiveFilterPushdown.java

```diff
@@ -0,0 +1,681 @@
+            extends ConnectorPlanRewriter<Void>
+    {
+        private final ConnectorSession session;
+        private final PlanNodeIdAllocator idAllocator;
+
+        public Rewriter(ConnectorSession session, PlanNodeIdAllocator idAllocator)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode filter, RewriteContext<Void> context)
+        {
+            if (!(filter.getSource() instanceof TableScanNode)) {
+                return visitPlan(filter, context);
+            }
+
+            TableScanNode tableScan = (TableScanNode) filter.getSource();
+            if (!isPushdownFilterSupported(session, tableScan.getTable())) {
+                return filter;
+            }
+
+            RowExpression expression = filter.getPredicate();
+            TableHandle handle = tableScan.getTable();
+            ConnectorMetadata metadata = getMetadata(handle);
+
+            BiMap<VariableReferenceExpression, VariableReferenceExpression> symbolToColumnMapping =
+                    getSymbolToColumnMapping(tableScan, handle, metadata, session);
+
+            RowExpression replacedExpression = replaceExpression(expression, symbolToColumnMapping);
+            // replaceExpression() may further optimize the expression;
+            // if the resulting expression is always false, then return empty Values node
+            if (FALSE_CONSTANT.equals(replacedExpression)) {
+                session.getWarningCollector().add(getPrestoWarning(tableScan));
+                return new ValuesNode(
+                        tableScan.getSourceLocation(),
+                        idAllocator.getNextId(),
+                        tableScan.getOutputVariables(),
+                        ImmutableList.of(),
+                        Optional.of(tableScan.getTable().getConnectorHandle().toString()));
+            }
+            ConnectorPushdownFilterResult pushdownFilterResult = pushdownFilter(
+                    session,
+                    metadata,
+                    handle.getConnectorHandle(),
+                    replacedExpression,
+                    handle.getLayout());
+
+            ConnectorTableLayout layout = pushdownFilterResult.getLayout();
+            if (layout.getPredicate().isNone()) {
+                session.getWarningCollector().add(getPrestoWarning(tableScan));
+                return new ValuesNode(
+                        tableScan.getSourceLocation(),
+                        idAllocator.getNextId(),
+                        tableScan.getOutputVariables(),
+                        ImmutableList.of(),
+                        Optional.of(tableScan.getTable().getConnectorHandle().toString()));
+            }
+
+            return getPlanNode(tableScan, handle, symbolToColumnMapping, pushdownFilterResult, layout, idAllocator);
```

Comment:
Remove getPlanNode() function and just call getTableScanNode(). Leave the handling of unenforcedFilter inline.